feat: Permit mapper plugins as valid blocks

ReubenFrankel opened the pull request:
<!--

Please, go through these steps when you submit a PR.

1. Make sure your branch is not protected. In particular, avoid making PRs from the `main` branch of your fork.

2. Give a descriptive title to your PR. We use semantic titles, and the accepted types and scopes are listed in https://github.com/meltano/meltano/blob/main/.github/semantic.yml.

   A good title should look like this:

   ```
   feat(cli): The `meltano run` command now accepts a `--timeout` option to limit the time it runs
   ```

3. Provide a description of your changes.

4. Put "Closes #XXXX" in your comment to auto-close the issue that your PR fixes (if such).

-->

## Description

Removes the

```block violates set requirements: Expected unique mappings name not the mapper plugin name:```

error message to optimistically allow mappers to be invoked directly, without having to instead reference a mapping.

As far as I see it, a mapping is just aliased config for mapper plugins. Some mappers will only ever need to perform a single function, so it doesn't makes sense to define a mapping in that case - you may as well just invoke the mapper directly; which is what this PR is addressing. For `meltano-map-transformer` and similar, I believe there is still value in mappings as it prevents you having to `inherit_from` a base `meltano-map-transformer` to define your reusable mappings (although I think they are functionally the same) - i.e.

```yml
plugins:
  mappers:
  - name: meltano-map-transformer
    variant: meltano
    pip_url: meltano-map-transform
    mappings:
    - name: all-caps
      config:
        stream_maps:
          '*':
            __alias__: __stream_name__.upper()
    - name: no-caps
      config:
        stream_maps:
          '*':
            __alias__: __stream_name__.lower()
```

and

```yml
plugins:
  mappers:
  - name: meltano-map-transformer
    variant: meltano
    pip_url: meltano-map-transform
  - name: all-caps
    inherit_from: meltano-map-transformer
    config:
      stream_maps:
        '*':
          __alias__: __stream_name__.upper()
  - name: no-caps
    inherit_from: meltano-map-transformer
    config:
      stream_maps:
        '*':
          __alias__: __stream_name__.lower()
```

## Related Issues

* Partially addresses #3233
* Related to #7992
* Related GitLab issues:
  * https://gitlab.com/meltano/meltano/-/merge_requests/2481
  * https://gitlab.com/meltano/meltano/-/issues/2299
  * https://gitlab.com/meltano/meltano/-/issues/3273

ReubenFrankel committed:
Permit mapper plugins as valid blocks

ReubenFrankel committed:
Merge branch 'main' into meltano-run-mapper

edgarrmondragon committed:
Merge branch 'main' into meltano-run-mapper

ReubenFrankel commented:
@edgarrmondragon Should we just remove the portion of the failing test that checks for an exception on mapper/mapping collision (https://github.com/meltano/meltano/actions/runs/13532529659/job/37817760752?pr=9105#step:12:2257), or is there something better we can do?

edgarrmondragon commented:
> @edgarrmondragon Should we just remove the portion of the failing `test_run_mapper_config` test that checks for an exception on mapper/mapping collision (https://github.com/meltano/meltano/actions/runs/13532529659/job/37817760752?pr=9105#step:12:2257), or is there something better we can do?

We probably just need to remove that test.

ReubenFrankel committed:
Remove part of test that expects an error when a mapper plugin with no mappings is invoked by the plugin name

ReubenFrankel committed:
Update expected error when a mapper plugin with a mapping of the same name is invoked by the ambiguous plugin/mapping name

ReubenFrankel commented:
> > @edgarrmondragon Should we just remove the portion of the failing `test_run_mapper_config` test that checks for an exception on mapper/mapping collision (https://github.com/meltano/meltano/actions/runs/13532529659/job/37817760752?pr=9105#step:12:2257), or is there something better we can do?
> 
> We probably just need to remove that test.

I removed the part of the test in 96fd5a405006da03353c7bdfcb56ebe6d8a92646. I updated a separate part in d4047a9aa96f794fc0c7c06cfdb16f4fbed2c834 to assert an `AmbiguousMappingName` error raised when a mapper plugin name and mapping name conflict, but couldn't find where to apply the change to meet that expectation. Unless you were talking about removing the whole test?

ReubenFrankel committed:
Merge branch 'main' into meltano-run-mapper

ReubenFrankel committed:
Merge branch 'main' into meltano-run-mapper

edgarrmondragon commented:
> Unless you were talking about removing the whole test?

I think this is definitely easier, so I'm OK with taking that path.

ReubenFrankel committed:
Remove `test_run_mapper_config` entirely

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.ci hooks

for more information, see https://pre-commit.ci

edgarrmondragon committed:
Merge branch 'main' into meltano-run-mapper

edgarrmondragon commented:
Trying to think what unexpected behavior this might cause for users, but can't think of any 🤔

ReubenFrankel commented:
> Trying to think what unexpected behavior this might cause for users, but can't think of any 🤔

There is no check for conflicts between mapper plugin name and mapping name. I don't actually know which one it is resolving here:

`meltano.yml`
```
  - name: mapper-fivetran
    variant: matatika
    pip_url: git+https://github.com/Matatika/mapper-fivetran.git
    mappings:
    - name: mapper-fivetran
```

```
meltano invoke mapper-fivetran
meltano run <tap> mapper-fivetran <target>
```

This is what I was talking about above:

edgarrmondragon committed:
Merge branch 'main' into meltano-run-mapper

ReubenFrankel committed:
Merge branch 'main' into meltano-run-mapper

edgarrmondragon committed:
Merge branch 'main' into meltano-run-mapper

edgarrmondragon committed:
Merge branch 'main' into meltano-run-mapper

edgarrmondragon committed:
Merge branch 'main' into meltano-run-mapper

edgarrmondragon committed:
Check ambiguous mapping and mapper name

edgarrmondragon committed:
Merge branch 'main' into meltano-run-mapper

edgarrmondragon committed:
Split exception classes

edgarrmondragon reviewed (changes requested):
The mapper-level `config` doesn't seem to be used and instead the _last_ found mapping's is used 🤔
